openstack/identity/identity.go

```diff
@@ -205,71 +205,68 @@ func validateProjectRole(client *gophercloud.ServiceClient, token string, projec
 	return false
 }
 
-func (h Handler) tenantToken(r *http.Request, tenant string) bool {
-	token := r.Header["X-Auth-Token"]
-	if token == nil {
-		return false
-	}
+func (h Handler) checkToken(r *http.Request, tenant string, token string) bool {
 
 	/* TODO Caching or PKI */
-	for _, s := range h.ValidServices {
-		if validateService(h.Client, token[0], tenant, s.ServiceType, s.ServiceName) == true {
+	for _, a := range h.ValidAdmins {
+		if validateProjectRole(h.Client, token, a.Project, a.Role) == true {
 			return true
 		}
-
 	}
 
 	for _, s := range h.ValidServices {
-		if validateService(h.Client, token[0], tenant, s.ServiceType, "") == true {
+		if validateService(h.Client, token, tenant, s.ServiceType, s.ServiceName) == true {
 			return true
 		}
 
 	}
 
-	return false
-}
-
-func (h Handler) adminToken(r *http.Request) bool {
-	token := r.Header["X-Auth-Token"]
-	if token == nil {
-		return false
-	}
-
-	/* TODO Caching or PKI */
-	for _, a := range h.ValidAdmins {
-		if validateProjectRole(h.Client, token[0], a.Project, a.Role) == true {
+	for _, s := range h.ValidServices {
+		if validateService(h.Client, token, tenant, s.ServiceType, "") == true {
```

Comment:
Could this be moved up into the prior loop now?  Then the h.ValidServices is only walked once instead of twice.